### ISSUE TITLE
Data race in pkg/kubelet/prober

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#18708]
When running with Go 1.5.1 there is the following race:

    Write by goroutine 81:
      k8s.io/kubernetes/pkg/kubelet/prober.TestHandleCrash()
          /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/worker_test.go:254 +0x42
      testing.tRunner()
          /tmp/workdir/go/src/testing/testing.go:456 +0xdc

    Previous read by goroutine 24:
      k8s.io/kubernetes/pkg/util.HandleCrash()
          /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:59 +0x34
      k8s.io/kubernetes/pkg/kubelet/prober.(*worker).doProbe()
          /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/worker.go:129 +0x409
      k8s.io/kubernetes/pkg/kubelet/prober.(*worker).run()
          /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/worker.go:109 +0xd6

@timstclair @yujuhong @kubernetes/goog-node 

This is one of blockers of upgrading to Go 1.5 - can you please take a look?

**Comment by timstclair** [comment]
Do you have a way to reliably reproduce this? Or do I just need to keep rerunning the tests until I hit it?

**Comment by timstclair** [comment]
Ah sorry, just realised I was missing the `-race` flag.

**Comment by wojtek-t** [comment]
Also - this was with Golang 1.5 - with that Go version it's failing mostly every run. And I believe in our default configuration we're running tests in parallel (--p=2 ?)

**Comment by madhusudancs** [comment]
Reopening this issue. Here are the test artifacts - http://pr-test.k8s.io/21131/kubernetes-pull-test-unit-integration/14156/

**Comment by roberthbailey** [comment]
Also saw this in http://pr-test.k8s.io/21131/kubernetes-pull-test-unit-integration/13727/

**Comment by bgrant0607** [comment]
And here:
https://00e9e64bac67da3a5483df52977e087fed44cbe55bb1d589f9-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/pr-logs%2Fpull%2F20557%2Fkubernetes-pull-test-unit-integration%2F14315%2Fartifacts%2Fjunit_v1,extensions-v1beta1_20160214-100306.stdout?qk=AD5uMEuawx7TUST9xxgNjq9kAyWmQF0S8phoAZ5LEJXmg51CxJGHFZQnkORCCmoxa7itNT3_0NSBW_PHfQBGHe535eXWUdXSa_EzknCjAMF221YRTIxm1xXDA6a102McgRGPVLsAkuR7yoBZJ0P0OEMklhoeVfPEXWb29XFTfiELTXNW1WHnSqm2FD7YSrljXGgL30uqvbsID5TM0RxKhLDJqacY7FeNMy82rsuQHezGEDj5P2HNNAv-blbcnNyAoZhtnzvnIPh0L4WlIzYUIse8r00SLckmsHdp9GjdTe9u1rGqchV5Bzkf7Mv0iq9yL4LtwsjEHMmpBu-puX5szcIEUz9OPvoKHnw62W-xK4EqFYM64gOLfwar-SsVIIBbrsae6eOsWgVl7JPdL7rjiBEieRgfb6-V5tZycCap-ypT2zGQEKIt2LpoE2AjbHdCpB9lT5dXNn0G3NRnvsFF9CWnz2gFg2u1MchfLVvuAhb1AEgAduNLVT2Gr1tIUZ0qBzsIGyl3AzrrKLozQf0yKZeFSrax0R9wTni7qnDopMGWkxcWpgq9RJ2V-ZEgAo8cwvXhbG1oZS0B-yR0isdYjoXB0GUpIVtcbXNPEhr5RyolQrIj3UPBjaxCaaOTaSOD2JYv3ul1ubgjvmT8zyqhia0scYHufcxjasCRie8TmwSKBhDGjG3HzIoZwGL1LiuMAmZIejAV8B9FUbKMakR7Pygc6_trD9ObJnvhSz6tq62ZdBi_-yKEPAb1rCQuI3a3omShAUswhLyfiegxfdoDJD01vr2TesTtukMO7toix8aI0NtgJH6ImsO9aQAHV-C4929xDyTlsy1zL6iB0jaZB1f7YVncChvYJ3t2voY_Z5O2lYmFdsQYkVpJy9f_B74bASVVg6EN6q0aKa6POFykam0MYT6GjujMqjAPiNowtpj5tS-s5VA9P6QGxW1KADv2vLmYFj5JDgX1-00VGjpPiigVV4Q6sUKVNJAiVumbJ7Z1p0C08JB2tEg

**Comment by bprashanth** [comment]
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/21326/kubernetes-pull-test-unit-integration/14794
```
=== RUN   TestUpdateReadiness
I0217 10:39:43.609019   21660 manager_test.go:386] Polling for ready state true
W0217 10:39:44.612828   21660 manager.go:211] PodStatus missing PodReady condition: {Phase:Running Conditions:[] Message: Reason: HostIP: PodIP: StartTime:2016-02-17 10:39:43.608819787 -0800 PST ContainerStatuses:[{Name:cOnTaInEr_NaMe State:{Waiting:<nil> Running:0xc820256b60 Terminated:<nil>} LastTerminationState:{Waiting:<nil> Running:<nil> Terminated:<nil>} Ready:true RestartCount:0 Image: ImageID: ContainerID:test://cOnTaInEr_Id}]}
==================
WARNING: DATA RACE
Write by goroutine 10:
  k8s.io/kubernetes/pkg/kubelet/status.normalizeStatus.func1()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/status/manager.go:457 +0x91
  k8s.io/kubernetes/pkg/kubelet/status.normalizeStatus.func2()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/status/manager.go:461 +0x80
  k8s.io/kubernetes/pkg/kubelet/status.normalizeStatus()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/status/manager.go:479 +0x2b8
  k8s.io/kubernetes/pkg/kubelet/status.isStatusEqual()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/status/manager.go:112 +0x32
  k8s.io/kubernetes/pkg/kubelet/status.(*manager).updateStatusInternal()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/status/manager.go:272 +0x456
  k8s.io/kubernetes/pkg/kubelet/status.(*manager).SetContainerReadiness()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/status/manager.go:215 +0x1385
  k8s.io/kubernetes/pkg/kubelet/prober.(*manager).updateReadiness()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/manager.go:237 +0x1c3
  k8s.io/kubernetes/pkg/kubelet/prober.(*manager).(k8s.io/kubernetes/pkg/kubelet/prober.updateReadiness)-fm()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/manager.go:100 +0x2d
  k8s.io/kubernetes/pkg/util/wait.JitterUntil.func1()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:66 +0x58
  k8s.io/kubernetes/pkg/util/wait.JitterUntil()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:67 +0x7d
  k8s.io/kubernetes/pkg/util/wait.Until()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:47 +0x4b

Previous read by goroutine 11:
  k8s.io/kubernetes/pkg/kubelet/prober.(*worker).doProbe()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/worker.go:181 +0x11d7
  k8s.io/kubernetes/pkg/kubelet/prober.(*worker).run()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/worker.go:115 +0x105

Goroutine 10 (running) created at:
  k8s.io/kubernetes/pkg/kubelet/prober.TestUpdateReadiness()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/manager_test.go:290 +0x3b4
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc

Goroutine 11 (running) created at:
  k8s.io/kubernetes/pkg/kubelet/prober.(*manager).AddPod()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/manager.go:147 +0x5c2
  k8s.io/kubernetes/pkg/kubelet/prober.TestUpdateReadiness()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/prober/manager_test.go:303 +0x5c9
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc
==================
```

**Comment by ixdy** [comment]
http://pr-test.k8s.io/21208/kubernetes-pull-test-unit-integration/14857/ shows the exact same data race as https://github.com/kubernetes/kubernetes/issues/18708#issuecomment-185371666.

**Comment by timstclair** [comment]
Any tips for reproducing this locally?

**Comment by bprashanth** [comment]
I think stress will help repro, given the frequency: https://github.com/kubernetes/kubernetes/blob/4ee1bd5b50fbb9ccecb567c1df835d0b85c1eb59/docs/devel/flaky-tests.md#reproducing-unit-test-flakes

Previous attempts of debugging hard to hit races I've had to upload a pr with logging and hammer the unittester over a weekend

**Comment by timstclair** [comment]
This should be fixed by #21448, please reopen if it happens again.
